jina/executors/indexers/keyvalue.py

```diff
@@ -67,14 +67,46 @@ def add(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
             self.write_handler.body.write(value)
             self._size += 1
             # print(f'l: {l} p: {p} r: {r} r+l: {r + l} size: {self._size}')
+        self.write_handler.flush()
 
     def query(self, key: int) -> Optional[bytes]:
+        if self.query_handler is None:
+            self.query_handler = self.get_query_handler()
         pos_info = self.query_handler.header.get(key, None)
         if pos_info is not None:
             p, r, l = pos_info
             with mmap.mmap(self.query_handler.body, offset=p, length=l) as m:
                 return m[r:]
 
+    def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        # check: hard fail (raises) on key not found
```

Comment:
@cristianmtr what about the sharding case, where we want to have an update on all shards having the key. I think, if  a shard don't have a key, it should silently ignore an update or delete request. I don't think there is a way of knowing which key is persistent in which shard from outside.

Reply:
I see what you mean. But isn't that something the tail pea should handle? It should catch these errors, right?

Reply:
okay, then let's leave it like this for now.